### PR TITLE
Kevin/kubernetes 1.8

### PR DESCRIPTION
**Pull request by knisbet**
Updates rigging to use Custom Resource Definitions on kubernetes 1.8 or newer, and continues to use third party resources on kubernetes 1.7 or lower.

I think Sasha mentioned in one of his emails, that it isn't necessary to support converting stored rigging entries from thirdpartyresources to CRDs, so I don't attempt to control for upgrades or do any migrations from third party resources to custom resources.

### CODE REVIEW

File: changeset.go [hunk 1]
```diff
@@ -1547,16 +1558,62 @@ func (cs *Changeset) upsertSecret(ctx context.Context, tr *ChangesetResource, da
...
+			}
+		}
+		// wait for the controller to init by trying to list stuff
+		return retry(ctx, 30, time.Second, func() error {
+			_, err := cs.list(DefaultNamespace)
+			return err
+		})
+	}
+	// kubernetes 1.8 or newer
+	crd := &apiextensions.CustomResourceDefinition{
```
Comment (a-palchikov) [inline changeset.go]: this too could be factored into `createCustomResourceDefinition`

File: Gopkg.toml [hunk 2]
```diff
@@ -44,5 +43,185 @@
...
 [[constraint]]
-  version = "4.0.0"
+  branch = "release-1.8"
+  name = "k8s.io/apiextensions-apiserver"
+
+[[constraint]]
+  version = "5.0.0"
   name = "k8s.io/client-go"
+
+[[override]]
```
[the remark this answers is not in the document]
Comment (knisbet) [reply Gopkg.toml]: As far as I could tell, this was the recommended way to have dep to enforce the version when vendoring sub packages. So this is a conversion of the client-go glide constraints into something dep can work with.

https://github.com/kubernetes/client-go/blob/master/INSTALL.md#dep-not-supported-yet

File: changeset.go [hunk 3]
```diff
@@ -1547,16 +1558,62 @@ func (cs *Changeset) upsertSecret(ctx context.Context, tr *ChangesetResource, da
...
+		return trace.Wrap(err)
+	}
+	versionMinor, err := strconv.Atoi(version.Minor)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// TODO(knisbet) deprecated. Remove Third Party resources when we're no longer supporting kubernetes < 1.8
+	if versionMajor == 1 && versionMinor < 8 {
+		tpr := &v1beta1.ThirdPartyResource{
```
[the remark this answers is not in the document]
Comment (knisbet) [reply changeset.go]: I'm not sure I agree this add's all that much value, keeping in mind that ThirdPartyResource is deprecated code. So when ThirdPartyResources get removed, the function will be back to only creating the CustomResourceDefinition.

[the remark this answers is not in the document]
Comment (a-palchikov) [reply changeset.go]: Ok - I'm fine with changes, so feel free to merge.

**Pull request merged**